keras_cv/layers/object_detection/roi_align.py

```diff
@@ -0,0 +1,367 @@
+        )
+
+        # Compute grid positions.
+        kernel_y, kernel_x, box_gridy0y1, box_gridx0x1 = _compute_grid_positions(
+            boxes, boundary, output_size, sample_offset
+        )
+
+        x_indices = tf.cast(
+            tf.reshape(box_gridx0x1, [batch_size, num_boxes, output_size * 2]),
+            dtype=tf.int32,
+        )
+        y_indices = tf.cast(
+            tf.reshape(box_gridy0y1, [batch_size, num_boxes, output_size * 2]),
+            dtype=tf.int32,
+        )
+
+        batch_size_offset = tf.tile(
+            tf.reshape(tf.range(batch_size) * batch_dim_size, [batch_size, 1, 1, 1]),
+            [1, num_boxes, output_size * 2, output_size * 2],
+        )
+        # Get level offset for each box. Each box belongs to one level.
+        levels_offset = tf.tile(
+            tf.reshape(
+                tf.gather(level_dim_offsets, levels), [batch_size, num_boxes, 1, 1]
+            ),
+            [1, 1, output_size * 2, output_size * 2],
+        )
+        y_indices_offset = tf.tile(
+            tf.reshape(
+                y_indices * tf.expand_dims(tf.gather(height_dim_sizes, levels), -1),
+                [batch_size, num_boxes, output_size * 2, 1],
+            ),
+            [1, 1, 1, output_size * 2],
+        )
+        x_indices_offset = tf.tile(
+            tf.reshape(x_indices, [batch_size, num_boxes, 1, output_size * 2]),
+            [1, 1, output_size * 2, 1],
+        )
+        indices = tf.reshape(
+            batch_size_offset + levels_offset + y_indices_offset + x_indices_offset,
+            [-1],
+        )
+
+        # TODO(tanzhenyu): replace tf.gather with tf.gather_nd and try to get similar
+        # performance.
+        features_per_box = tf.reshape(
+            tf.gather(features_r2, indices),
+            [batch_size, num_boxes, output_size * 2, output_size * 2, num_filters],
+        )
+
+        # Bilinear interpolation.
+        features_per_box = _feature_bilinear_interpolation(
+            features_per_box, kernel_y, kernel_x
+        )
+        return features_per_box
+
+
+class _ROIAligner(tf.keras.layers.Layer):
+    """Performs ROIAlign for the second stage processing."""
+
+    def __init__(
+        self, bounding_box_format, target_size=7, sample_offset: float = 0.5, **kwargs
+    ):
+        """Initializes a ROI aligner.
+        Args:
+          bounding_box_format: the input format for boxes.
+          crop_size: An `int` of the output size of the cropped features.
+          sample_offset: A `float` in [0, 1] of the subpixel sample offset.
+          **kwargs: Additional keyword arguments passed to Layer.
+        """
+        self._config_dict = {
+            "bounding_box_format": bounding_box_format,
+            "crop_size": target_size,
+            "sample_offset": sample_offset,
+        }
+        super(_ROIAligner, self).__init__(**kwargs)
+
+    def call(
+        self, features: Mapping[str, tf.Tensor], boxes: tf.Tensor, training: bool = None
```

Comment:
I like the annotation for features and boxes; but

is a type annotation here useful for `training`?  Normally, I wouldn't comment on something so small; but it actually looks incorrect even:

```
training:bool = None
```

breaks the bool contract

Reply:
Yeah, should be Optional.
Done